Lib/test/test_iter.py

```diff
@@ -238,6 +244,52 @@ def test_mutating_seq_class_exhausted_iter(self):
         self.assertEqual(list(empit), [5, 6])
         self.assertEqual(list(a), [0, 1, 2, 3, 4, 5, 6])
 
+    def test_reduce_mutating_builtins_iter(self):
+        # This is a reproducer of issue #101765
+        # where iter `__reduce__` calls could lead to a segfault or SystemError
+        # depending on the order of C argument evaluation, which is undefined
+
+        # Backup builtins.iter
+        builtins = __builtins__.__dict__ if hasattr(__builtins__, "__dict__") else __builtins__
```

Comment:
```suggestion
        builtins = getattr(__builtins__, "__dict__", __builtins__)
```

in what cases (that need to be supported here) does `__builtins__` not have a `__dict__`?

Reply:
https://docs.python.org/3/reference/executionmodel.html?highlight=__builtins__#builtins-and-restricted-execution `__builtins__` is already a dict when not in the `__main__` module. I was thinking we may have both cases but I suppose we tests only run with the non `__main__` case, or in other words we only need `builtins = __builtins__`?

Reply:
Switched to `builtins = __builtins__` in [178b8ea](https://github.com/python/cpython/pull/101769/commits/178b8ea41bb045eed63d06674e2f337d4de97bef) since I don't see how unittest can be a `__main__` module

Reply:
The simplest and clearest approach might be to `import builtins` as recommended in the docs; then it will always be a module.

Reply:
changed to `import builtins` in [98ec3c6](https://github.com/python/cpython/pull/101769/commits/98ec3c63a973d06e784f8a04bcb67d8c324808ec)